main/src/ca/uwaterloo/flix/language/ast/Type.scala

```diff
@@ -160,26 +160,85 @@ sealed trait Type {
     case _ => throw InternalCompilerException(s"Unexpected non-arrow type: '$this'.")
   }
 
+  /**
+    * Returns the argument types of `this` curried arrow type.
+    * Returns `Nil` if `this` is not an arrow type.
+    *
+    * For example,
+    *
+    * {{{
+    * Int32                               =>     Nil
+    * Int32 -> String -> Int32            =>     List(Int32, String)
+    * (Int32, String) -> String -> Bool   =>     List(Int32, String, String)
+    * }}}
+    */
+  def curriedArrowArgTypes: List[Type] = typeConstructor match {
+    case Some(TypeConstructor.Arrow(_)) => arrowArgTypes ++ arrowResultType.curriedArrowArgTypes
+    case _ => Nil
+  }
+
   /**
     * Returns the result type of `this` arrow type.
     *
     * NB: Assumes that `this` type is an arrow.
     */
   def arrowResultType: Type = typeConstructor match {
-    case Some(TypeConstructor.Arrow(n)) => typeArguments.last
+    case Some(TypeConstructor.Arrow(_)) => typeArguments.last
     case _ => throw InternalCompilerException(s"Unexpected non-arrow type: '$this'.")
   }
 
+  /**
+    * Returns the result type of `this` curried arrow type.
+    *
+    * For example,
+    *
+    * {{{
+    * Int32                               =>     throw
+    * Int32 -> String -> Int32            =>     Int32
+    * (Int32, String) -> String -> Bool   =>     Bool
+    * }}}
+    *
+    * NB: Assumes that `this` type is an arrow.
+    */
+  def curriedArrowResultType: Type = {
+    val resType = arrowResultType
+    resType.typeConstructor match {
+      case Some(TypeConstructor.Arrow(_)) => resType.curriedArrowResultType
+      case _ => resType
+    }
+  }
+
   /**
     * Returns the effect type of `this` arrow type.
     *
     * NB: Assumes that `this` type is an arrow.
     */
   def arrowEffectType: Type = typeConstructor match {
-    case Some(TypeConstructor.Arrow(n)) => typeArguments.head
+    case Some(TypeConstructor.Arrow(_)) => typeArguments.head
     case _ => throw InternalCompilerException(s"Unexpected non-arrow type: '$this'.")
   }
 
+  /**
+    * Returns the effect type of `this` curried arrow type.
+    *
+    * For example,
+    *
+    * {{{
+    * Int32                                        =>     throw
+    * Int32 -> String -> Int32 & Pure              =>     Pure
+    * (Int32, String) -> String -> Bool & Impure   =>     Impure
+    * }}}
+    *
+    * NB: Assumes that `this` type is an arrow.
+    */
+  def curriedArrowEffectType: Type = {
+    val resType = arrowResultType
+    resType.typeConstructor match {
+      case Some(TypeConstructor.Arrow(_)) => resType.curriedArrowEffectType
+      case _ => arrowEffectType
+    }
+  }
+
```

Comment:
I understand that you need these functions. It also makes sense to put them on Type.

However, I would prefer to move them to where they are actually used. 

The reason is that they are quite special and brittle, and we *do not* other parts of the compiler to depend on them.

Reply:
In other words. Type tries hard to only expose a few core functions that a widely used.